### PR TITLE
Add Distributed Trace Propagation

### PR DESCRIPTION
**Pull request by ian-mi**
Enables trace propagation using the distributed tracing extension by default in the eventing client with a new option to disable trace propagation. Adds a HasTracePropagation transport property which transports can use to indicate that the transport's incoming trace context should be preferred to the context defined by the distributed tracing extension.

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
I'm not sure it's a good idea to add a method about extensions to the `Transport` interface, nor add extension logic to the `Client` code directly... Should we reason about creating some abstractions to decouple this code?

**Comment by ian-mi** [comment]
Could you elaborate on what type of abstraction you feel is lacking? The alternative to handling this in the client seemed to be adding this capability into each transport which would involve duplicating code across each transport. Client seemed like the appropriate place to put cross-cutting functionality but the client needs some information about the capabilities of the transport.

**Comment by n3wscott** [comment]
@slinkydeveloper  This is adding the official CloudEvents trace extension to the SDK. We are suppose to support official extensions from https://github.com/cloudevents/spec/blob/master/SDK.md#extensions.

File: pkg/cloudevents/transport/transport.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ type Transport interface {
 	SetConverter(Converter)
 	// HasConverter is true when a non-nil converter has been set.
 	HasConverter() bool
+
+	HasTracePropagation() bool
```
Comment (n3wscott) [inline pkg/cloudevents/transport/transport.go]: Can you add godoc to this?

File: pkg/cloudevents/transport/http/transport.go [hunk 2]
```diff
@@ -353,6 +353,12 @@ func (t *Transport) StartReceiver(ctx context.Context) error {
 	}
 }
 
+// HasTracePropagation implements Transport.HasTracePropagation
+// TODO: implement HTTP trace propagation.
```
Comment (n3wscott) [inline pkg/cloudevents/transport/http/transport.go]: this todo means there is still more to enable this feature?

File: pkg/cloudevents/client/client.go [hunk 3]
```diff
@@ -107,6 +111,15 @@ func (c *ceClient) obsSend(ctx context.Context, event cloudevents.Event) (contex
 		}
 	}
 
+	// Set distributed tracing extension.
+	if !c.disableTracePropagation {
+		if span := trace.FromContext(ctx); span != nil {
```
Comment (n3wscott) [inline pkg/cloudevents/client/client.go]: So spans are only propagated if they exist on the context, correct?

**Comment by n3wscott** [comment]
What is the plan with regards to `tracestate` ? https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md#tracestate

File: pkg/cloudevents/transport/http/transport.go (hunk 2 above)
Comment (ian-mi) [reply pkg/cloudevents/transport/http/transport.go]: I've gone ahead and added HTTP propagation in this PR.

File: pkg/cloudevents/client/client.go (hunk 3 above)
Comment (ian-mi) [reply pkg/cloudevents/client/client.go]: Correct, although if tracing is enabled in the SDK a new trace will be created immediately before this.

**Comment by ian-mi** [comment]
This PR should correctly handle tracestate. I've made parsing fail-open however so an invalid tracestate will not fail a request and will not prevent a valid traceparent from being used.

**Comment by n3wscott** [comment]
LGTM

**Pull request merged**